### DISCUSSION TITLE
Separate CNN model for individual digit(s)

### DISCUSSION THREAD

**Discussion by niobos** [jomjol/AI-on-the-edge-device#1481] [Ideas] [answered]
### The Feature

Hi,

First of all, thank you for this wonderful project. I really appreciate the effort and time you put in to this. Thank you!

I have a feature request/idea. I'm not very familiar with AI/CNNs, so it maybe it's stupid. In that case: please explain why, so I can learn!

Currently you need to select a CNN for digit recognition, and a second CNN for analog needle recognition. I think it would be beneficial to be able to specify a separate CNN per digit. In my case, I'd use a class11-CNN for most of the digits, but use a separately trained class100-CNN for the last digit. I think it would improve the recognition rate of both style of digits.

If you agree this looks like a good idea. I'm willing to give it a try to implement it. I have some ESP-IDF experience, but I'm not sure how far that will get me on this project.

**Comment by caco3** [comment]
do you get bad readings if you use the same model for all digits?
wouldn't it be easier to just train the models better?

can you provide a screenshot?

**Comment by 8OND007** [comment]
same here, I use your fantastic project several months now on my electrical ferrarris counter (spinning wheels). 

For the first 6 digital (whole) I would like to keep using the dig-class11 since these number have to be considered as whole number (4 not 3.9 for example).
Only the last digit in my counter has be be recognized with dig-class100 (internmediate states) since this is a spinning number (red) 0,1 to 0,9. (1 digit behind the decimal point/comma)

Problem with using the dig-class100 for all numbers is that it sometimes recogizes false intermediate states in hgher decimal ranges. (hundreds, thousands,...)

For ex. my real counter is 9 9 8 6 7 8 . 5
Sometimes the 6 digit (hundreds) is slightly higher or lower turned, but in fact has to be recognized as 6
Only the last digit 5 (behind the comma/point) has to be recoginzed with intermediate states.
If I use the dig-class100 intermediate detection is on ALL numbers, so the slightly false turned number 6 is recognized as 5.9 or 6.1 which gives a false high deviation on the hundreds number.

As suggested by niobos it would be great to allow per digit it's recognition algo.
So I set all digits to dig-class11 and only the last last digit to dig-class100.

I just started to use the dig_class11 now to see if it's better, before that I used the dig-s2-q-20220211 (see available list below):

1. dig-s2-q-20220211
2. dig1330s1q
3. dig_class100-140_s2_q
4. dig_cont_0600_s3
5. dig-class11_1500_s2

I don't see any description on the wiki page of numbers 1, 2 en 4 ?
What is difference between number 1 en 5 ?

**Comment by niobos** [comment]
My suggestion comes from efficiency: Why make it harder for the CNN to detect 100 classes if I know there will only be 10? It will inevitable reduce the accuracy of the CNN, and make it slower as well.

I haven't tried running the class100 on all digits, since I'm still struggling to get accurate readings with a (self-trained) class11. Being able to combine class11 and class100 would also allow me to gradually get training data for the class100 digit (the least significant one), while keeping good accuracy on the class11 digits (all but the least significant one).

**Comment by jomjol** [comment, marked as answer]
@niobos : even if you only have 10 number, if it is not a LCD, then the intermediate values help the post processing values to judge on the correct value.

@all: 
* I believe, the better way is to incorporate wrong detected images as additional training data into the existing network. With this it will get more and more robust for a wide range of meters.
* There I will not work on a config option to enable different cnn files for each digit. If this is a which, you would need to make a fork with this feature.

I will convert this issue to a discussion, there we can continue a discussion.

**Comment by 8OND007** [comment]
ok, just out of curiousity I tried the dig-class100-0140_s2_q.tfite
 instead of the dig-class1.
As suspected it recognized turned hundred, thousands, .... also as 3.8 or 4.2 instead of 4, but it the final calculation it rounds the number to whole ? Since the send counter on MQTT is correct.
![meterAI](https://user-images.githubusercontent.com/43314247/205511047-952ee64e-0720-438e-b36f-44c71cf870cd.png)
![meterAI2](https://user-images.githubusercontent.com/43314247/205511054-b2c5ee70-2d38-43ee-9f21-76ddae065496.png)

So I think i'm going to let is stay globally on dig-class100-0140_s2_q.tfite
I also hope the AI now "filters" bad spikes (up or down) out. Until now I sometimes had small bumps/spikes which mess up home assistant's energy usage with dig-class11. Hopefully this will be flattend out now with the intermediate states (and not us N anymore)

![meter_spikes](https://user-images.githubusercontent.com/43314247/205511225-d480c579-71a3-4420-b341-ee6d3c120043.png)

**Reply by jomjol** [reply]
It does not only round trivial, it uses the subdigit information for consistency changes.
